bug/pciu-service-no-method-error

hpjaj opened the pull request:
## Background

We started getting these errors in Sentry logging for production two days ago:

```
NoMethodErrorlib/common/client/middleware/request/immutable_headers.rb in call at line 66
```

Coming from this sentry issue:

http://sentry.vetsgov-internal/vets-gov/platform-api-production/issues/37304/

Here is an associated issue in the Faraday repo:  

https://github.com/lostisland/faraday/issues/727

As it turns out, under the hood, the file `net/http/header.rb` calls `value = value.strip`.  If `value` is `nil`, this errors with `NoMethodError: undefined method 'strip' for nil:NilClass`.

## Definition of done

- [x] Set any `nil` header values to `''`
- [x] Leave any non-`nil` header values as is
- [x] Leave empty header hashes empty
- [x] Spec coverage

rlbaker commented on lib/common/client/base.rb:
```diff
@@ -67,6 +68,14 @@ def request(method, path, params = {}, headers = {})
+        headers.transform_values! do |value|
+          value.nil? ? '' : value
+        end
+      end
```
Could this be simplified using map?

```
def sanitize_headers!(headers)
  headers.map do |k, v|
    [k.to_s, v.nil? ? '' : v]
  end
end
```

rlbaker replied:
After talking this through with harry, this would require other changes to support the non-mutating nature of map.

hpjaj commented:
@kreek - In order to try this out in dev and staging, before hitting prod, I added a conditional around a setting that will only render true in dev and staging, not in prod.

This key will remain false for the next few weeks.

As this is a hot fix, this is the quickest way to get feedback in dev and staging, and not impact prod.

I also added sentry logging to collect data.

Anything else we should add for protective measures and data collection?

cc: @rlbaker

kreek commented on lib/common/client/base.rb:
```diff
@@ -45,7 +45,16 @@ def perform(method, path, params, headers = nil)
+        log_message_to_sentry(
+          'nil headers bug',
+          :info,
+          headers: headers, method: method, path: path, params: params, client: inspect,
```
I'd check that Sentry obfuscates all PII, EVSS and other systems send it along in headers

hpjaj replied:
Good call.  Completed in https://github.com/department-of-veterans-affairs/vets-api/pull/1894/commits/e7dc63ee37b7af8128af5b35bd28083276abb160

kreek commented:
I like testing it for a while in staging, are we more worried about `sanitize_headers!` throwing an error or that the resulting call would fail? (or both). If it's only the former you could add a rescue block within `sanitize_headers!` that, if it goes sideways, logs an error and returns the original unmodified headers.

hpjaj commented:
@kreek - Yeah, concerned with both.  Moreover, want to tread lightly here, b/c of the importance of this method, and how much traffic goes through it.

As such, since we only need to log when the bug pops up, which is when one of the header’s value are `nil`, I favored only logging to sentry at that point.  At least for now, until proven we need to add more logging.

How does that sound?

kreek commented on lib/common/client/base.rb:
```diff
@@ -67,6 +68,28 @@ def request(method, path, params = {}, headers = {})
+              log_message_to_sentry(
+                'nil headers bug',
+                :info,
+                unmodified_headers: unmodified_headers, method: method, path: path, params: params, client: inspect,
```
could get some double logging here as it logs all the headers for each one that's nil

hpjaj replied:
Sorry, not following you.  I removed the sentry logging from the `request` method itself, and moved it into this `sanitize` method.  So we should only get logging once per request, and only for the `nil` case.

Or perhaps I am missing something (which is likely the case)?
